Avoid exception on getting the focus target of compound widgets

level420 opened the pull request:
Focus able widgets may not have a "textfield" child control resulting in an exception calling getChildControl("textfield"). To avoid this we first have to check if a child control named "textfield" exists.

level420 committed:
Avoid exception on getting the focus target of compund widgets

Focus able widgets may not have a "textfield" child control resulting in an exception calling getChildControl("textfield"). To avoid this we first have to check if a child control named "textfield" exists.

level420 commented:
@ronkie could you please have a look into this PR? Thank you!

johnspackman commented on framework/source/class/qx/event/handler/Focus.js:
```diff
@@ -1056,11 +1056,12 @@ qx.Class.define("qx.event.handler.Focus",
-          return textField.getContentElement().getDomElement();
+        var widget = qx.ui.core.Widget.getWidgetByElement(focusedElement);
+        if(widget) {
+          var textField = widget.getChildControl && widget.hasChildControl("textfield") && widget.getChildControl("textfield", true);
```
as we're sniffing by method name, then it's necessary to sniff for `hasChildControl` also - there is no contract that says that `hasChildControl` will exist if `getChildControl` does.

level420 replied:
Hmm. As of the api docs `qx.ui.core.Widget.getWidgetByElement` allways returns an instance of `qx.ui.core.Widget` which always should implement `hasChildControl` and `getChildControl`.
I think we don't need the checks at all.

level420 committed:
remove unnecessary check for existence of method getChildControl

as qx.ui.core.Widget.getWidgetByElement should always return instances of `qx.ui.core.Widget` or their subclasses, which have both `hasChildControl` and `getChildControl`

ronkie reviewed (approved):
The changes are fine with me. However, as far as I see, getChildControl method does not throw anything itself. Only _createChildControl can throw an exception when control is not supported, but we don't call it when calling getChildControl with "notcreate == true".

In this case we also have a check for textField on line 1062, which seems enough to me.

level420 commented:
Yep! Found one of my classes where I've overridden `getChildControl` without passing the second optional parameter `notcreate`. I'm unsure what I should do now with this PR. Close it?

derrell commented:
If it turns out that the PR is unnecessary, then yes, I'd say close it.

level420 commented:
Closed because this was an error in my own code.
